Newsletters: add 237 (2023-02-08)

harding opened the pull request:
- [x] Add releases/RCs and topic links @harding
- [x] Bitcoin Core PR Review Club summary @LarryRuane
- [x] Bitcoin Core 25880 @Xekyo
- [x] Core Lightning 5679 @adamjonas

jonatack commented on _posts/en/newsletters/2023-02-08-newsletter.md:
```diff
@@ -0,0 +1,130 @@
+
+- [Core Lightning #5821][] adds `preapproveinvoice` (pre-approve
+  invoice) and `preapprovekeysend` (pre-approve keysend) RPCs that allow
+  the caller to send either a [BOLT11][] invoice or a [keysend][topic
```
"to send either a BOLT11 invoice or a keysend payment details to" ... drop "a" here?

jonatack commented on _posts/en/newsletters/2023-02-08-newsletter.md:
```diff
@@ -0,0 +1,130 @@
+  the caller to send either a [BOLT11][] invoice or a [keysend][topic
+  spontaneous payments] payment details to Core Lightning's signing module
+  (`hsmd`) to verify the module is willing to sign the payment.  For
+  some applications, such as those where the amount of money which can
```
```suggestion
  some applications, such as those where the amount of money that can
```

(I believe "which" would require a preceding comma ---  "money, which" --- as done in line 105 here below)

jonatack commented on _posts/en/newsletters/2023-02-08-newsletter.md:
```diff
@@ -0,0 +1,130 @@
+  partner at funding time to only use the closing address specified at
+  that time (see [Newsletter #158][news158 upfront] for details).
+
+- [LND #7252][] adds support for using Sqlite as LND's database backend.
```
```suggestion
- [LND #7252][] adds support for using SQLite as LND's database backend.
```

harding replied:
I opened a for-discussion PR about updating the style guide to have us capitalize proper nouns using standard English rules, see https://github.com/bitcoinops/bitcoinops.github.io/pull/1005 ; for now, I suggest we keep this text as-is.  If we decide against the proposed policy, I'll come back and update this string.

jonatack commented on _posts/en/newsletters/2023-02-08-newsletter.md:
```diff
@@ -0,0 +1,130 @@
+  problems than simply attempting the payment and dealing with failure.
+
+- [Core Lightning #5849][] makes backend changes that allow a node to
+  handle over 100,000 peers each with one channel.  Although someone
```
I wasn't sure I was grokking correctly, looked at the PR, and "quickly open 100,000 peers with one channel each" in the merged pull description seems clearer to me.

jonatack commented on _posts/en/newsletters/2023-02-08-newsletter.md:
```diff
@@ -0,0 +1,130 @@
+    being stored in other ways.  Those other ways might raise costs for
+    those generating the data---but probably not by enough to
+    significantly discourage the behavior---and the alternative storage
+    methods might create new problems for traditional Bitcoin users.
```
We're beginning to see ML proposals for storing this data offchain, like "Purely off-chain coin colouring'" https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2023-February/021396.html by Anthony Towns.

jonatack reviewed (commented):
Initial read-through

mzumsande commented on _posts/en/newsletters/2023-02-08-newsletter.md:
```diff
@@ -54,12 +54,80 @@ Bitcoin infrastructure software.
+      in `AddrMan`, finding no outbound peers? How does this PR fix it?"
+  a2="If the `-onlynet` configuration changes. For example, suppose the
+      node has always been run with `-onlynet=ipv4` so its AddMan has
+      no ipv6 addresses. Then the node is restarted with `-onlynet=ipv6`.
```
I would replace ipv6 with onion or i2p in this example. Switching between clearnet and a privacy network makes much more sense as a use case.

jonatack replied:
> I would replace ipv6 with onion or i2p in this example. Switching between clearnet and a privacy network makes much more sense as a use case.

FWIW I believe neither of these would be typical real-world cases and that most onlynet users are either setting onlynet=onion, or since a couple of years now, also onlynet=onion and i2p.

LarryRuane replied:
Oh, just saw your comment here moments after my latest push... Is it worth fixing, do you think?

jonatack replied:
I think it's more credible to use a real world case but it's just a suggestion.

mzumsande replied:
Not sure if this discussion belongs here - but I think it's quite possible that a user would do IBD with a node in default clearnet mode (without setting up tor, so that Tor is unreachable), and after finishing IBD, they would install/activate tor and decide to switch to `-onlynet=tor`. This is now possible without requiring an intermediate period in which they'd need to be connected to multiple networks or requiring any manual steps, like deleting `peers.dat` or setting up manual connections.

jonatack replied:
Agree. Though once the node has peers in its reachable networks, manual connections are likely the best way to ensure it remains connected to at least one honest peer, more so with Tor and I2P if it's easier for an attacker to spin up many nodes.  I should probably update doc/{tor,i2p,cjdns}.md (and my blog) to communicate this.

LarryRuane replied:
> I think it's quite possible that a user would do IBD with a node in default clearnet mode (without setting up tor, so that Tor is unreachable), and after finishing IBD, they would install/activate tor and decide to switch to `-onlynet=onion`.

Just as a point of information, that's _exactly_ with I did with the little RPi [myNode](https://mynodebtc.com/) when I bought it last year. I started the IBD over default onion, and it was going so slowly that it was going to take months (rough estimate). I switched to clearnet and it finished in about a week (still slow, but it's a tiny CPU). Then I switched back to the default onion. I probably had some peers in addrman left over from the first phase? I didn't notice it getting stuck when I switched back to onion.

mzumsande commented on _posts/en/newsletters/2023-02-08-newsletter.md:
```diff
@@ -54,12 +54,80 @@ Bitcoin infrastructure software.
+[Track AddrMan totals by network and table, improve precision of adding fixed seeds][review club 26847]
+is a PR by Martin Zumsande, co-authored by Amiti Uttarwar, that
+allows the Bitcoin Core client to more reliably find outbound peers
+and distribute them more broadly across available network types (for
```
would skip the "distribute" part, this PR doesn't touch address relay (how addresses are distributed).

LarryRuane replied:
Ah, yes, I was using "distribute" in a different sense, not in the sense of distributing addresses, but the distribution of our peers (more specifically, their network types). This will be a common confusion, fixed.

mzumsande commented on _posts/en/newsletters/2023-02-08-newsletter.md:
```diff
@@ -54,12 +54,80 @@ Bitcoin infrastructure software.
-FIXME is a PR by FIXME that FIXME:LarryRuane
+[Track AddrMan totals by network and table, improve precision of adding fixed seeds][review club 26847]
+is a PR by Martin Zumsande, co-authored by Amiti Uttarwar, that
+allows the Bitcoin Core client to more reliably find outbound peers
```
maybe a bit weaker, "... in certain situations" or so.
The fix in this PR really only improves the `-onlynet` situation described in Q2, not finding outbound peers in general.

mzumsande commented on _posts/en/newsletters/2023-02-08-newsletter.md:
```diff
@@ -54,12 +54,80 @@ Bitcoin infrastructure software.
+  a5="Nothing in the PR _guarantees_ any particular distribution of
+      peers among the reachable networks. For example, if we had 10k
+      clearnet addresses and only 1k I2P addresses in `AddrMan`, it's
+      likely that all our peers will be clearnet (ipv4 or ipv6)."
```
would use a much smaller number of I2P addresses in the example (maybe 50 or 100). Otherwise, with 10 outbound peers, the chance to draw at least one peer at the given ratio would actually be significant.

LarryRuane replied:
Yes, I actually thought of this while writing this part. But I was trying to stay true to the IRC chat log. But I agree, it's probably better to give an accurate answer here, fixed.

jonatack replied:
If helpful, here are some real-world numbers from my node at the moment for the alternative networks. The clearnet numbers are relatively low because recently the node has been running onlynet=onion/i2p/cjdns and so the number of non-IsTerrible clearnet peers gradually drops.

<img width="477" alt="Screenshot 2023-02-03 at 14 16 33" src="https://user-images.githubusercontent.com/2415484/216721334-2ef8c5c7-2edf-44aa-832d-27617e6b7a7f.png">

paulkania commented on _posts/en/newsletters/2023-02-08-newsletter.md:
```diff
@@ -0,0 +1,130 @@
+
+- **Discussion about storing data in the block chain:** users of a new
+  project recently began storing large amounts of data in the witness
+  data for segwit v1 ([taproot][topic taproot]) transactions.   Robert
```
why is the link wrapped in parentheses in this case?

LarryRuane replied:
To me, that looks reasonable, since it's saying segwit v1, and taproot is just another name for segwit v1 (to help anyone who might not be aware of that fact).

paulkania replied:
Ah, I thought it was for markup reasons.
i'm trying to figure out why my PR [#1001](https://github.com/bitcoinops/bitcoinops.github.io/pull/1001) isn't passing CI, and it has something to do with [text][topic x] not converting to html properly.

LarryRuane reviewed (commented):
Thanks, Martin, all comments addressed (7e7371f915c98b846184d09346ecd38e193d0c3a)

jonatack commented on _posts/en/newsletters/2023-02-08-newsletter.md:
```diff
@@ -0,0 +1,199 @@
+      together, avoiding chain splits caused by network partitions.
+      If half the nodes, including miners, ran with `-onlynet=x` and
+      the other half, including miners, ran `-onlynet=y`, then two
+      chains could emerge."
```
Would note that these positions are gradually changing.  There are currently around 14k Tor peers and 1k I2P peers.  Also, from discussions with node operators and as recommended by Bitcoin Core documentation, those who are using onlynet are increasingly doing so with multiple networks.

LarryRuane replied:
That's interesting, but I think a little beyond the point I was trying to make here (about possible chain splits). Also, I don't want to stray too far from what was discussed during the review club meeting (this is supposed to be just a summary of the meeting).

jonatack replied:
It should probably also be noted that it's possible to have an outbound connection to each reachable network at all times simply by adding manual connections to your conf file.

LarryRuane replied:
That's really good to know, and I hadn't thought of that myself, but again I'm not sure it's appropriate to add things here that weren't covered in the review club meeting at all. If it _is_ okay to do that, let me know, and I'd be glad to add it!

harding replied:
It's ok to add things that weren't covered during the meeting if they make the material more useful.

LarryRuane commented:
Pushed another commit, fac2a230daca5e590d84b48b0340c9ca3748c26e, to address review comments (thanks, @jonatack!). Feel free to squash. (When multiple people are pushing to the same branch, I believe it's better not to do force-pushes.)

paulkania reviewed (commented):
review complete: no suggestions, great job.

Question: Why are most [links][topic x] colored purple, while a few do not, e.g. output linking Line 102, hardware wallet Line  71  ?

harding commented:
@LarryRuane 

> Feel free to squash.

FYI, @bitschmidty kindly squashes everything down to a single commit per author at the end, so there's no need to do that during the editing process.

>When multiple people are pushing to the same branch, I believe it's better not to do force-pushes.

You may want to read the docs for `git push --force-with-lease` which won't push unless your copy of the remote branch has the same tip as the remote branch itself during the force push operation.  E.g., if you fetch this branch, then I push a commit, and then you try to push with a lease, your push will fail.

harding commented:
Made edits for feedback, except any comments that might remain in @LarryRuane's section.  Thanks @jonatack @mzumsande @LarryRuane @paulkania for comments!  Added topic links.  Dropped releases/RC section (none for the third week running, wow!).  Reviewed contributions by @adamjonas and @LarryRuane (thanks!).  Added Bitcoin Core 25880 and a sentence saying we'll provide summaries of the additional Ordinals/Inscriptions mailing list discussions in next week's newsletter.

Thanks everyone!

Xekyo commented:
Ah, sorry, I was late. I was just trying to push when I noticed that you wrote a description.

- [Bitcoin Core #25880][] makes the stalling timeout adaptive during the 
    initial synchronization. Bitcoin Core requests blocks from multiple
    peers in parallel. If one peer is significantly slower than others to
    the point that the node gets stuck waiting for the next block, we
    disconnect the stalling peer after a timeout. In some situations this
    could cause a node with a low-bandwidth connection to disconnect
    multiple peers in a row when a heavy block could not be transferred
    within the timeout. This code change amends the behavior of nodes to
    dynamically adapt the timeout: the timeout is incremented for each
    disconnected peer while no block is received, and after blocks start
    arriving again, the timeout is scaled back block-by-block.

mzumsande commented on _posts/en/newsletters/2023-02-08-newsletter.md:
```diff
@@ -141,7 +142,21 @@ Interface (HWI)][hwi repo], [Rust Bitcoin][rust bitcoin repo], [BTCPay
-- [Bitcoin Core #25880][] p2p: Make stalling timeout adaptive during IBD FIXME:Xekyo
+- [Bitcoin Core #25880][] fixes a problem where a node could be unable
+  to acquire additional blocks during Initial Block chain Download
+  (IBD).  Previously, if the node responsible for delivering the next
```
The 2s timeout is not a global timeout for each block, but only a "last resort" deadline if the node has gotten itself into a stalling situation. The specifics for how that can happen (1024 moving block window exhausted etc.)  are probably out the scope for this text, but one could misunderstand this such that peers only have 2s to provide a block as a general rule.

harding replied:
Updating the text to @Xekyo's which provides the extra context.

vostrnad commented on _posts/en/newsletters/2023-02-08-newsletter.md:
```diff
@@ -0,0 +1,217 @@
+
+- **Discussion about storing data in the block chain:** users of a new
+  project recently began storing large amounts of data in the witness
+  data for segwit v1 ([taproot][topic taproot]) transactions.   Robert
```
```suggestion
  data for segwit v1 ([taproot][topic taproot]) inputs.  Robert
```
"segwit v1" is a really a property of an input/output, not transaction

vostrnad commented on _posts/en/newsletters/2023-02-08-newsletter.md:
```diff
@@ -0,0 +1,217 @@
+  a0="A network is assumed to be reachable unless we're sure we
+      can't access it, or our configuration has specified one or
+      more _other_ networks using the `-onlynet=` configuration option
+      (only those are considered reachable, even if other network
```
```suggestion
      (only those are then considered reachable, even if other network
```

vostrnad commented on _posts/en/newsletters/2023-02-08-newsletter.md:
```diff
@@ -0,0 +1,217 @@
+Proposals (BIPs)][bips repo], and [Lightning BOLTs][bolts repo].*
+
+- [Bitcoin Core #25880][] fixes a problem where a node could be unable
+  to acquire additional blocks during Initial Block chain Download
```
```suggestion
  to acquire additional blocks during Initial Block Download
```
(as per STYLE.md)

harding commented:
Pushed updates, thanks @Xekyo @vostrnad and @mzumsande !

LarryRuane commented:
> > When multiple people are pushing to the same branch, I believe it's better not to do force-pushes.
> 
> You may want to read the docs for `git push --force-with-lease` which won't push unless your copy of the remote branch has the same tip as the remote branch itself during the force push operation. E.g., if you fetch this branch, then I push a commit, and then you try to push with a lease, your push will fail.

Thanks, that's great to know! But from what you're saying, it's best if we all just make separate little commits (not force-pushes), and then let @bitschmidty squash at the very end. (LMK if that's incorrect.)

LarryRuane commented:
> Made edits for feedback, except any comments that might remain in @LarryRuane's section.

I'm making a few small fixes at this very moment, hope it's not too late, I'll be done in a few minutes...

LarryRuane commented:
Added a new commit 81cdb4764845a534d6a61b7fc68db7f8a0946d80 to incorporate https://github.com/bitcoinops/bitcoinops.github.io/pull/999#discussion_r1096349473, thanks @jonatack!

bitschmidty commented:
Pushed a few small styling fixups to the PR Review section and squashed.

bitschmidty commented:
Thank you @harding @LarryRuane @adamjonas and @Xekyo for authoring this week and @jonatack @mzumsande @vostrnad @paulkania for reviewing!
